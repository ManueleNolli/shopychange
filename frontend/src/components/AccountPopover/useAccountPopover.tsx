import { useColorModeValue } from '@chakra-ui/react'

export default function useAccountPopover() {
  const popoverContentBgColor = useColorModeValue('white', 'gray.800')

  return { popoverContentBgColor }
}
