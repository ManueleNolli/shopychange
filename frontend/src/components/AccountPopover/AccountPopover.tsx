import React from 'react'

import { AccountIconProps } from '../../types/props/AccountIconProps'
import { AccountNavItem } from '../../types/components/AccountNavItem'
import { Link as LinkRouter } from 'react-router-dom'

import { colors } from '../../styles/Colors'

import {
  Box,
  Button,
  Popover,
  PopoverContent,
  PopoverTrigger,
  Stack,
  Link,
  useColorModeValue,
  Flex,
  Center,
  Text,
} from '@chakra-ui/react'
import { FaUser } from 'react-icons/fa'
import useAccountPopover from './useAccountPopover'

export default function AccountPopover({ navItems }: AccountIconProps) {
  const { popoverContentBgColor } = useAccountPopover()

  return (
    <Box>
      <Popover placement={'bottom-start'}>
        <PopoverTrigger>
          <Button
            data-testid="account-button"
            aria-label="Account"
            _focus={{ boxShadow: 'none' }}
            w="fit-content"
          >
            <FaUser />
          </Button>
        </PopoverTrigger>
        {navItems ? (
          <PopoverContent
            border={0}
            boxShadow={'xl'}
            bg={popoverContentBgColor}
            p={4}
            rounded={'xl'}
            w={{ base: '40vw', md: '20vw' }}
            maxW={{ base: '200px', md: '200px' }}
          >
            <Stack>
              {navItems.map((item) => (
                <Item key={item.label} {...item} />
              ))}
            </Stack>
          </PopoverContent>
        ) : null}
      </Popover>
    </Box>
  )
}

const Item = ({ label, icon, href, action }: AccountNavItem) => {
  if (action) {
    return (
      <Box
        onClick={action}
        // change hover mouse cursor
        _hover={{
          cursor: 'pointer',
        }}
      >
        <ItemContent label={label} icon={icon} />
      </Box>
    )
  } else {
    return (
      <Link as={LinkRouter} to={href ?? '#'} style={{ textDecoration: 'none' }}>
        <ItemContent label={label} icon={icon} />
      </Link>
    )
  }
}

const ItemContent = ({ label, icon }: AccountNavItem) => {
  return (
    <Box
      p={4}
      role={'group'}
      rounded={'md'}
      _hover={{
        bg: useColorModeValue(`${colors.popover.bgChakraColor}.50`, 'gray.900'),
        color: `${colors.popover.bgChakraColor}.400`,
      }}
      transition={'color .3s ease, transform .3s ease'}
    >
      <Flex>
        <Center
          flex="4"
          _groupHover={{ transform: 'scale(1.33)' }}
          transition={'transform .3s ease'}
          data-testid={`icon-${label}`}
        >
          {icon}
        </Center>
        <Box flex="6">
          <Text fontSize={'sm'} fontWeight={500} data-testid={`label-${label}`}>
            {label}
          </Text>
        </Box>
      </Flex>
    </Box>
  )
}
