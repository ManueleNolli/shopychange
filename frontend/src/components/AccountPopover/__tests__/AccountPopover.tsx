import React from 'react'
import { render, screen } from '@testing-library/react'
import AccountPopover from '../AccountPopover'
import { BrowserRouter } from 'react-router-dom'

import { FaUser } from 'react-icons/fa'

const actionMock = jest.fn()

const navItems = [
  {
    label: 'Profile',
    icon: <FaUser />,
    href: '/account',
  },
  {
    label: 'Watchlist',
    icon: <FaUser />,
    href: '/account/watchlist',
  },
  {
    label: 'Create',
    icon: <FaUser />,
    href: '/account/create',
  },
  {
    label: 'Language',
    icon: <FaUser />,
    href: '/account/personalize',
  },
  {
    label: 'Disconnect',
    icon: <FaUser />,
    action: actionMock,
  },
]

describe('AccountPopover', () => {
  test('renders correctly', () => {
    const tree = render(<AccountPopover navItems={navItems} />, {
      wrapper: BrowserRouter,
    })
    expect(tree.container).toMatchSnapshot()

    expect(screen.getByTestId('label-Profile')).toBeInTheDocument()
    expect(screen.getByTestId('label-Watchlist')).toBeInTheDocument()
    expect(screen.getByTestId('label-Create')).toBeInTheDocument()
    expect(screen.getByTestId('label-Language')).toBeInTheDocument()
    expect(screen.getByTestId('label-Disconnect')).toBeInTheDocument()

    // test action button
    const disconnectButton = screen.getByTestId('label-Disconnect')
    expect(disconnectButton).toBeInTheDocument()
    disconnectButton.click()
    expect(actionMock).toHaveBeenCalled()
  })
})
