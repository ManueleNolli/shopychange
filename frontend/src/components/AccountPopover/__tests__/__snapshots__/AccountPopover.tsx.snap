// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountPopover renders correctly 1`] = `
<div>
  <div
    class="css-0"
  >
    <button
      aria-controls="popover-content-:r1:"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Account"
      class="chakra-button css-1ke86mo"
      data-testid="account-button"
      id="popover-trigger-:r1:"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
        />
      </svg>
    </button>
    <div
      class="chakra-popover__popper css-0"
      style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
    >
      <section
        class="chakra-popover__content css-fvx4v9"
        id="popover-content-:r1:"
        role="dialog"
        style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
        tabindex="-1"
      >
        <div
          class="chakra-stack css-8g8ihq"
        >
          <a
            class="chakra-link css-0"
            href="/account"
            style="text-decoration: none;"
          >
            <div
              class="css-10tm5g0"
              role="group"
            >
              <div
                class="css-k008qs"
              >
                <div
                  class="css-avkhbc"
                  data-testid="icon-Profile"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                    />
                  </svg>
                </div>
                <div
                  class="css-1vou5f1"
                >
                  <p
                    class="chakra-text css-wu0fx1"
                    data-testid="label-Profile"
                  >
                    Profile
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            class="chakra-link css-0"
            href="/account/watchlist"
            style="text-decoration: none;"
          >
            <div
              class="css-10tm5g0"
              role="group"
            >
              <div
                class="css-k008qs"
              >
                <div
                  class="css-avkhbc"
                  data-testid="icon-Watchlist"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                    />
                  </svg>
                </div>
                <div
                  class="css-1vou5f1"
                >
                  <p
                    class="chakra-text css-wu0fx1"
                    data-testid="label-Watchlist"
                  >
                    Watchlist
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            class="chakra-link css-0"
            href="/account/create"
            style="text-decoration: none;"
          >
            <div
              class="css-10tm5g0"
              role="group"
            >
              <div
                class="css-k008qs"
              >
                <div
                  class="css-avkhbc"
                  data-testid="icon-Create"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                    />
                  </svg>
                </div>
                <div
                  class="css-1vou5f1"
                >
                  <p
                    class="chakra-text css-wu0fx1"
                    data-testid="label-Create"
                  >
                    Create
                  </p>
                </div>
              </div>
            </div>
          </a>
          <a
            class="chakra-link css-0"
            href="/account/personalize"
            style="text-decoration: none;"
          >
            <div
              class="css-10tm5g0"
              role="group"
            >
              <div
                class="css-k008qs"
              >
                <div
                  class="css-avkhbc"
                  data-testid="icon-Language"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                    />
                  </svg>
                </div>
                <div
                  class="css-1vou5f1"
                >
                  <p
                    class="chakra-text css-wu0fx1"
                    data-testid="label-Language"
                  >
                    Language
                  </p>
                </div>
              </div>
            </div>
          </a>
          <div
            class="css-1m8iww1"
          >
            <div
              class="css-10tm5g0"
              role="group"
            >
              <div
                class="css-k008qs"
              >
                <div
                  class="css-avkhbc"
                  data-testid="icon-Disconnect"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                    />
                  </svg>
                </div>
                <div
                  class="css-1vou5f1"
                >
                  <p
                    class="chakra-text css-wu0fx1"
                    data-testid="label-Disconnect"
                  >
                    Disconnect
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;
